synapse/handlers/federation.py

```diff
@@ -2903,21 +2915,25 @@ async def _check_key_revocation(self, public_key, url):
 
     async def persist_events_and_notify(
         self,
+        room_id: str,
         event_and_contexts: Sequence[Tuple[EventBase, EventContext]],
         backfilled: bool = False,
     ) -> int:
         """Persists events and tells the notifier/pushers about them, if
         necessary.
 
         Args:
+            room_id:
             event_and_contexts:
             backfilled: Whether these events are a result of
                 backfilling or not
         """
-        if self.config.worker.writers.events != self._instance_name:
+        instance = self.config.worker.events_shard_config.get_instance(room_id)
+        if instance != self._instance_name:
```

Comment:
Edit: I wrote the below before realizing that `instance` is also used in the `_send_events` call below, I thought it was worth leaving though in case it jiggles a good idea loose:

> It looks like this pattern is used quite a bit, in the comments for `ShardedWorkerHandlingConfig` it says to prefer `should_handle`, which seems like it could be used here:
> 
> ```python
> if self.config.worker.events_shard_config.should_handle(self._instance_name, room_id):
> ```
> 
> Although I think some of this could be simplified more if `ShardedWorkerHandlingConfig` knew what the *current instance* was (maybe `should_handle` wouldn't need the `instance` passed in?)

Reply:
*nods*. The problem with giving `ShardedWorkerHandlingConfig` the current instance name is that I don't think we know what the instance name is during config parsing (outside of the worker config parsing).

Reply:
Actually, I think we do need to use `should_handle` first.

Reply:
Though I guess the fact that we go on to do a HTTP replication hit means that `get_instance` has to work.

Reply:
I think it is OK to use `get_instance` here, that and `should_handle` *should* have matching logic after all!